File: python/printer.py
import mal_types as types

def escape(str):
    return str.replace('\\', '\\\\')\
              .replace('"', '\\"')\
              .replace('\n', '\\n')

def pr_str(obj, print_readably=True):
    if type(obj) == list:
        return "(" + " ".join(map(lambda x: pr_str(x, print_readably), obj)) + ")"
    elif type(obj) == str:
        if print_readably:
            return '"' + escape(obj) + '"'
        else:
            return obj
    elif obj is None:
        return "nil"
    elif obj is True:
        return "true"
    elif obj is False:
        return "false"
    elif type(obj) == types.Atom:
        return "(atom " + pr_str(obj.val, print_readably) + ")"
    else:
        return obj.__str__()

File: js/printer.js
var types = require('./types')

function escape(str) {
    return str.replace(/\\/g, "\\\\")
              .replace(/"/g, '\\"')
              .replace(/\n/g, "\\n")
}

function pr_str(obj, print_readably) {
    if (obj instanceof types.Symbol) {
        return obj.name
    } else if (Array.isArray(obj)) {
        return "(" + obj.map(function(x) { return pr_str(x, print_readably) }).join(" ") + ")"
    } else if (typeof obj === 'string') {
        if (print_readably) {
            return '"' + escape(obj) + '"'
        } else {
            return obj
        }
    } else if (obj === null) {
        return 'nil'
    } else if (obj instanceof types.Atom) {
        return "(atom " + obj.value + ")"
    } else {
        return obj.toString()
    }
}

exports.pr_str = pr_str

